### PR TITLE
Improve escaping of byte, byte str, and c str proc-macro literals

### PR DESCRIPTION
**Pull request by dtolnay**
This PR changes the behavior of `proc_macro::Literal::byte_character` (https://github.com/rust-lang/rust/issues/115268), `byte_string`, and `c_string` (https://github.com/rust-lang/rust/issues/119750) to improve their choice of escape sequences. 3 categories of changes are made:

1. Never use `\x00`. Always prefer `\0`, which is supported in all the same places.

2. Never escape `\'` inside double quotes and `\"` inside single quotes.

3. Never use `\x` for valid UTF-8 in literals that permit `\u`.

The second commit adds tests covering these cases, asserting the **old** behavior.

The third commit implements the behavior change and simultaneously updates the tests to assert the **new** behavior.

### CODE REVIEW

File: library/proc_macro/src/lib.rs [hunk 1]
```diff
@@ -34,6 +34,7 @@
 #![feature(rustc_attrs)]
 #![feature(min_specialization)]
 #![feature(strict_provenance)]
+#![feature(utf8_chunks)]
```
Comment (fee1-dead) [inline library/proc_macro/src/lib.rs:37]: It says above:

```
// This library is copied into rust-analyzer to allow loading rustc compiled proc macros.
// Please avoid unstable features where possible to minimize the amount of changes necessary
// to make it compile with rust-analyzer on stable.
```

Can we do escaping without using unstable features?

**Comment by petrochenkov** [comment]
I remember we talked about something related in https://github.com/rust-lang/rust/pull/95343, but don't remember details.
cc https://github.com/rust-lang/rust/issues/60495 as well.

File: library/proc_macro/src/lib.rs (hunk 1 above)
Comment (dtolnay) [reply library/proc_macro/src/lib.rs]: Good catch. I've changed this to use stable code only.

**Comment by fee1-dead** [comment]
Given [the FCP here](https://github.com/rust-lang/rust/issues/99543#issuecomment-2050455515) and that this is mostly an improvement / not a bug fix I'd say we should wait until it gets stabilized (on beta?)

**Comment by dtolnay** [comment]
If anything, having an official non-`core::str::Utf8Chunks` implementation checked in is an advantage because I'll need to use that in proc-macro2. Proc-macro2 is not going to be able to adopt `Utf8Chunks` as soon as it goes into beta.

**Comment by fee1-dead** [comment]
I don't think we should add burden of maintainability to a first-party library to ease the burden of maintainability of a third-party library.

**Comment by dtolnay** [comment]
@rustbot ready

The new implementation compiles with stable Rust 1.79, which contains the stabilization of `Utf8Chunks`.

**Review by fee1-dead — APPROVED** [review]
LGTM, thanks!

**Pull request merged**